File: week1_baseline/ruby/02_the_registry/lib/boukensha/registry.rb
require_relative "tool"
require_relative "errors"

module Boukensha
  class Registry
    def initialize(context)
      @context = context
    end

    def tool(name, description:, parameters: {}, &block)
      @context.register_tool(Tool.new(name, description, parameters, block))
    end

    def dispatch(name, args = {})
      tool = @context.tools[name]
      raise UnknownToolError, "No tool registered as '#{name}'" unless tool

      symbol_args = args.transform_keys(&:to_sym)
      tool.block.call(**symbol_args)
    end
  end
end

File: week1_baseline/python/06_the_logger/boukensha/registry.py
from typing import Any, Callable, Dict, Optional

from .errors import UnknownToolError
from .tool import Tool


class Registry:
    def __init__(self, context) -> None:
        self._context = context

    def tool(
        self,
        name: str,
        description: str,
        parameters: Optional[Dict[str, Any]] = None,
        block: Optional[Callable[..., Any]] = None,
    ) -> Tool:
        tool = Tool(name, description, parameters or {}, block)
        self._context.register_tool(tool)
        return tool

    def dispatch(self, name: str, args: Optional[Dict[str, Any]] = None) -> Any:
        tool = self._context.tools.get(name)
        if not tool:
            raise UnknownToolError("No tool registered as '{}'".format(name))

        return tool.block(**(args or {}))
